### ISSUE TITLE
warn for nonsensical use of @volatile

### ISSUE THREAD

**Issue by adriaanm** [scala/scala-dev#190]
the only time this makes sense is on a `var`, but it currently does not warn when added to a `def` (note that we should keep allowing it on trait vars, even though the annotation doesn't take effect until the field is synthesized in a subclass)

**Comment by jmpage** [comment]
A lot of things that don't make sense can be annotated with `@volatile`:

```
scala> @volatile type TEST = Int
defined type alias TEST

scala> @volatile object TestObj
defined object TestObj

scala> @volatile class TestCls
defined class TestCls
```

Would it make sense to make this an error since it's likely that someone who makes this mistake might intend for it to work but be surprised when it does not (and may not notice the warning until it is too late)?

**Comment by Ichoran** [comment]
There is all sorts of nonsensical stuff that you can do with annotations:

```scala
scala> @specialized val x = 7
x: Int = 7

scala> @annotation.tailrec type TEST = Int
defined type alias TEST

scala> @annotation.meta.getter trait Foo {}
defined trait Foo

scala> @annotation.switch class Bippy {}
defined class Bippy
```
What is the general policy on warning or raising errors on annotations?  Right now they seem to go on anything whether it makes any sense or not.

**Comment by adriaanm** [comment]
Yep, annotations are not very strictly enforced. `@volatile` is kind of weird, because we do warn in one specific case, but not in many others. I guess you could argue we should just not warn at all and silently drop it like we do for other non-sensical combos... For those annotations that influence bytecode, I do think we should be more helpful with our warnings.
